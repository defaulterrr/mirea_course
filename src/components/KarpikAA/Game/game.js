import React from 'react';
import './game_style.css';

class Game extends React.Component {
  sounds = {
    green: new Audio('https://s3.amazonaws.com/freecodecamp/simonSound1.mp3'),
    red: new Audio('https://s3.amazonaws.com/freecodecamp/simonSound2.mp3'),
    yellow: new Audio('https://s3.amazonaws.com/freecodecamp/simonSound3.mp3'),
    blue: new Audio('https://s3.amazonaws.com/freecodecamp/simonSound4.mp3'),
  };

  buttons = ['green', 'red', 'yellow', 'blue'];

  constructor(props) {
    super(props);
    this.state = {
      index: 0,
      round: 0,
      steps: [],
      text: '',
    };
  }

  playSound = (color) => {
    this.sounds[color].play();
    if (color === this.state.steps[this.state.index]) {
      this.setState({ index: this.index + 1 }, () => {
        if (this.state.index === this.state.round) {
          this.setState({
            index: 0,
          });
          this.generateStep();
          this.showSteps();
        }
      });
    } else if (this.state.round !== 0) {
      this.setState({
        text: `Game over! You reached round ${this.round}`,
      });
      this.resetSimon();
    }
  };

  // Generate a random color
  generateStep = () => {
    this.state.steps.push(this.buttons[(Math.floor(Math.random() * 4))]);
    this.setState({
      round: this.round + 1,
    });
  }

  // Play current step
  playStep = (step) => {
    const buttons = this.refs[step];
    buttons.classList.add('hover');
    this.sounds[step].play();
    setTimeout(
      () => {
        buttons.classList.remove('hover');
      },
      200,
    );
  };

  // Show all steps
  showSteps = () => {
    let num = 0;
    const moves = setInterval(
      () => {
        setTimeout((this.playStep(this.state.steps[num])), 300);
        this.setState({ text: 'Wait...' });
        num += 1;
        if (num >= this.state.steps.length) {
          this.setState({ text: 'Repeat the steps!' });
          clearInterval(moves);
        }
      },
      600,
    );
  };

  // Reset game
  resetSimon = (start) => {
    this.setState(
      {
        round: 0,
        index: 0,
        steps: [],
      },
      () => {
        if (start) {
          this.generateStep();
          this.showSteps();
        }
      },
    );
  };

  render() {
    return (
      <div>
        <h1>Simon Says</h1>
        <div className="simon">
          <div className="circle">
            <div className="inner">
              <div
                ref="green"
                className="buttons"
                id={this.buttons[0]}
                onClick={() => {}}
                onKeyPress={this.playSound(this.buttons[0])}
              />
              <div
                ref="red"
                className="buttons"
                id={this.buttons[1]}
                onClick={() => {}}
                onKeyPress={this.playSound(this.buttons[1])}
              />
              <div
                ref="yellow"
                className="buttons"
                id={this.buttons[2]}
                onClick={() => {}}
                onKeyPress={this.playSound(this.buttons[2])}
              />
              <div
                ref="blue"
                className="buttons"
                id={this.buttons[3]}
                onClick={() => {}}
                onKeyPress={this.playSound(this.buttons[3])}
              />
              <div className="round">{this.state.round}</div>
            </div>
          </div>
        </div>
        <div className="controls">
          <div className="text">{this.state.text}</div>
          <button type="submit" id="start" onClick={() => this.resetSimon(1)}>
            {!this.state.round ? 'Start Game' : 'Reset Game'}
          </button>
        </div>
      </div>
    );
  }
}

export default Game;
