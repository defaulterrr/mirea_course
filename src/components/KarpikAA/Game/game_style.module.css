* {
    box-sizing: border-box
}

*:focus {
    outline: none
}

h1 {
    color: #36383f;
    font-size: 30px;
	text-align: center;
}

.simon {
    position: relative;
    width: 400px;
    height: 400px;
    margin: 0 auto;
}

.circle {
    position: absolute;
    width: 400px;
    height: 400px;
    background-color: #36383f;
    border-radius: 50%;
    box-shadow: 0 0 5px black;
    border-bottom: 1px solid black;
}

.inner {
    margin: 30px;
}

.buttons {
    float: left;
    width: 150px;
    height: 150px;
    margin: 10px;
	cursor: pointer;
}

.green {
    float: left;
    width: 150px;
    height: 150px;
    margin: 10px;
	cursor: pointer;
    border-radius: 100% 3% 0 3%;
    background-color: #1ecd34;
}

.red {
    float: left;
    width: 150px;
    height: 150px;
    margin: 10px;
	cursor: pointer;
    border-radius: 3% 100% 3% 0;
    background-color: #fe3e2b;
}

.yellow {
    float: left;
    width: 150px;
    height: 150px;
    margin: 10px;
	cursor: pointer;
    border-radius: 3% 0 3% 100%;
    background-color: #fedd00;
}

.blue {
    float: left;
    width: 150px;
    height: 150px;
    margin: 10px;
	cursor: pointer;
    border-radius: 0 3% 100% 3%;
    background-color: #0070db;
}

.green:hover,
.red:hover,
.yellow:hover,
.blue:hover,
.hover {
    box-shadow: 0 0 30px #979caf;
}

.green:active,
.red:active,
.yellow:active,
.blue:active {
    box-shadow: 0 0 15px #585b66;
    opacity: 0.9;
}

.controls {
    width: 600px;
    margin: 0 auto;
    text-align: center;
}

.text {
    padding: 15px;
	color: #36383f;
    border-radius: 5px;
    margin: 30px;
}

#start {
    border: 0;
    background-color: #36383f;
    padding: 15px 30px;
    color: #fff;
	font-weight: 700;
	box-shadow: 0 0 1px black;
    border-radius: 5px;
	cursor: pointer;
	
}

#start:hover {
    box-shadow: 0 0 0 black;
}

#start:active {
    background-color: #2c2e34;
}

