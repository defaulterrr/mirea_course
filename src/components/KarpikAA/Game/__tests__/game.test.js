import React from 'react';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import Game from '../game';

configure({ adapter: new Adapter() });

describe('Render Game', () => {
  it('Render correctly', () => {
    const GameComponent = renderer.create(<Game />).toJSON();
    expect(GameComponent).toMatchSnapshot();
  });

  it('Should have zero states in the start', () => {
    const wrapper = shallow(<Game />);
    expect(wrapper.state('index')).toEqual(0);
    expect(wrapper.state('round')).toEqual(0);
    expect(wrapper.state('text')).toHaveLength(0);
    expect(wrapper.state('steps')).toHaveLength(0);
  });

  it('Should push steps after click', () => {
    const wrapper = shallow(<Game />);
    wrapper.find('#start').simulate('click');
    expect(wrapper.state('round')).toEqual(1);
    expect(wrapper.state('steps')).toHaveLength(1);
  });

  it('PlayStep should be called ones', () => {
    const playStep = jest.fn();
    setTimeout((playStep(2)), 300);
    expect(playStep).toHaveBeenCalledTimes(1);
  });

  it('Should play sound of red button with the click', () => {
    const app = shallow(<Game />);
    const playSoundSpy = jest.spyOn(app.instance(), 'playSound');
    app.update();
    app.instance().forceUpdate();
    const button = app.find('#red');
    button.simulate('click');
    expect(playSoundSpy).toHaveBeenCalled();
  });

  it('Should play sound of green button with the click', () => {
    const app = shallow(<Game />);
    const playSoundSpy = jest.spyOn(app.instance(), 'playSound');
    app.update();
    app.instance().forceUpdate();
    const button = app.find('#green');
    button.simulate('click');
    expect(playSoundSpy).toHaveBeenCalled();
  });

  it('Should play sound of yellow button with the click', () => {
    const app = shallow(<Game />);
    const playSoundSpy = jest.spyOn(app.instance(), 'playSound');
    app.update();
    app.instance().forceUpdate();
    const button = app.find('#yellow');
    button.simulate('click');
    expect(playSoundSpy).toHaveBeenCalled();
  });

  it('Should play sound of blue button with the click', () => {
    const app = shallow(<Game />);
    const playSoundSpy = jest.spyOn(app.instance(), 'playSound');
    app.update();
    app.instance().forceUpdate();
    const button = app.find('#blue');
    button.simulate('click');
    expect(playSoundSpy).toHaveBeenCalled();
  });

  it('Testing reset game', () => {
    const wrapper = shallow(<Game />);
    wrapper.find('#start').simulate('click');
    expect(wrapper.state('round')).toEqual(1);
    expect(wrapper.state('index')).toEqual(0);
    expect(wrapper.state('steps')).toHaveLength(1);

    wrapper.find('#start').simulate('click');
    expect(wrapper.state('round')).toEqual(1);
    expect(wrapper.state('index')).toEqual(0);
    expect(wrapper.state('steps')).toHaveLength(1);
  });

  it('Should call resetSimon after click on Start button', () => {
    const app = shallow(<Game />);
    const resetSimonSpy = jest.spyOn(app.instance(), 'resetSimon');
    app.update();
    app.instance().forceUpdate();
    const button = app.find('#start');
    button.simulate('click');
    expect(resetSimonSpy).toHaveBeenCalled();
  });
});
