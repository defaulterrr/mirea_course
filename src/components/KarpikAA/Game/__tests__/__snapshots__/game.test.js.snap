// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Game Render correctly 1`] = `
<div>
  <h1>
    Simon Says
  </h1>
  <div
    className="simon"
  >
    <div
      className="circle"
    >
      <div
        className="inner"
      >
        <div
          className="buttons"
          id="green"
          onClick={[Function]}
        />
        <div
          className="buttons"
          id="red"
          onClick={[Function]}
        />
        <div
          className="buttons"
          id="yellow"
          onClick={[Function]}
        />
        <div
          className="buttons"
          id="blue"
          onClick={[Function]}
        />
        <div
          className="round"
        >
          0
        </div>
      </div>
    </div>
  </div>
  <div
    className="controls"
  >
    <div
      className="text"
    >
      
    </div>
    <button
      id="start"
      onClick={[Function]}
      type="button"
    >
      Start Game
    </button>
  </div>
</div>
`;
